runtime/bcutil/cfreader.c

```diff
@@ -1783,11 +1799,18 @@ checkMethods(J9PortLibrary* portLib, J9CfrClassFile* classfile, U_8* segment, U_
 		} 
 
 		if (nameIndexOK && utf8Equal(&classfile->constantPool[method->nameIndex], "<init>", 6)) {
+			BOOLEAN classfileIsValuetype = J9_IS_CLASSFILE_VALUETYPE(classfile);
+#ifdef J9VM_OPT_VALHALLA_VALUE_TYPES
+#ifdef J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD
+			if (classfileIsValuetype) {
+				errorCode = J9NLS_CFR_ERR_INIT_ON_VALUE_CLASS__ID;
+				goto _errorFound;
+			}
+#endif /* #ifdef J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD */
+#endif /* #ifdef J9VM_OPT_VALHALLA_VALUE_TYPES */
 
 			/* check no invalid flags set */
-			if ((J9_IS_CLASSFILE_VALUETYPE(classfile) && (value & ~CFR_INIT_VT_METHOD_ACCESS_MASK))
```

Comment:
oops, I forgot the case for when `J9VM_OPT_VALHALLA_VALUE_TYPES` is defined but not `J9VM_OPT_VALHALLA_NEW_FACTORY_METHOD`. For the case when they are both defined, I am only checking `classfileIsValuetype` because in that case value classes are not allowed to have methods named `<init>` at all.

Reply:
changed in 3d60f49

Reply:
I think we should also do a check for `~CFR_INIT_VT_METHOD_ACCESS_MASK` for `<new>`.

Reply:
alright its changed in d6916e9